Source/Dafny/Verifier/Translator.TrStatement.cs

```diff
@@ -1381,12 +1383,12 @@ public partial class Translator {
         var ss = TrSplitExpr(loopInv.E, etran, false, out splitHappened);
         if (!splitHappened) {
           var wInv = Bpl.Expr.Imp(w, etran.TrExpr(loopInv.E));
-          invariants.Add(Assert(loopInv.E.tok, wInv, errorMessage ?? "loop invariant violation"));
+          invariants.Add(Assert(loopInv.E.tok, wInv, new DafnyLoopInvariantDescription(errorMessage)));
         } else {
           foreach (var split in ss) {
             var wInv = Bpl.Expr.Binary(split.E.tok, BinaryOperator.Opcode.Imp, w, split.E);
             if (split.IsChecked) {
-              invariants.Add(Assert(split.E.tok, wInv, errorMessage ?? "loop invariant violation"));  // TODO: it would be fine to have this use {:subsumption 0}
+              invariants.Add(Assert(split.E.tok, wInv, new DafnyLoopInvariantDescription(errorMessage)));  // TODO: it would be fine to have this use {:subsumption 0}
```

Comment:
Will it make a difference that we pass a non-null value in all cases now?

Reply:
The check for null has moved inside the `DafnyLoopInvariantDescription` class, and the non-null object being created here is stored in a different property than the string was before. It's a little fiddly, but I'm hoping the logic for null vs non-null strings goes away entirely before long.